apps/finance/contracts/Finance.sol

```diff
@@ -566,15 +600,15 @@ contract Finance is EtherTokenConstant, IsContract, AragonApp {
         }
     }
 
-    function _executePayment(uint256 _paymentId) internal {
+    function _executePayment(uint256 _paymentId) internal returns (uint256) {
```

Comment:
We could add an extra return value that is true if the `PaymentFailure` event was emitted, which would remove the need to perform a redundant `_nextPaymentTime(_paymentId) > getTimestamp64()` to decide the revert reason.

Reply:
Was looking at this and I don't think this covers the right cases. `PaymentFailure` is only emitted if at least one payment was due, but the funds weren't available, and needs the `_nextPaymentTime(_paymentId) > getTimestamp64()` in each iteration.

In the check, we want to detect if the payment wasn't due at all for deciding the revert reason.

Reply:
Might not be getting the full picture and this is a very small optimization, but the `PaymentFailure` event can only be emitted if there was a payment due that couldn't be performed. If we returned one extra value (`bool lastPaymentFailed`) then in `_executePaymentAtLeastOnce` we could just decide on the revert reason depending on `lastPaymentFailed` rather than re-checking if there was a payment due.

Reply:
Yes, I don't get it either. Assuming `paid == 0` back in `_executePaymentAtLeastOnce`, if `_nextPaymentTime(_paymentId) > getTimestamp64()` then it means that it didn't get into the loop of `_executePayment`, so `PaymentFailure` wasn't emitted; while if the opposite holds true, then it means it got into the loop and therefore it went through the `break` and so the event was emitted, right? Maybe I'm missing something too.

Reply:
Ohh yes, I see what @izqui is saying now. Yes, if `paid == 0` and a `PaymentFailure()` event is emitted then we will always be throwing `ERROR_EXECUTE_PAYMENT_NUM`.

I've reordered the `if`s in `_executePaymentAtLeastOnce()` to make this a bit more clear, but decided against the extra return value as it'd make successful calls more expensive.